core/trino-main/src/main/java/io/trino/execution/scheduler/SourcePartitionedScheduler.java

```diff
@@ -519,42 +529,55 @@ public void close()
             if (noMoreSplitsNotification.containsKey(node)) {
                 noMoreSplits.putAll(partitionedNode, noMoreSplitsNotification.get(node));
             }
-            newTasks.addAll(stage.scheduleSplits(
-                    node,
-                    splits,
-                    noMoreSplits.build()));
+            RemoteTask task = scheduledTasks.get(node);
+            if (task != null) {
+                task.addSplits(splits);
+                noMoreSplits.build().forEach(task::noMoreSplits);
+            }
+            else {
+                scheduleTask(node, splits, noMoreSplits.build()).ifPresent(newTasks::add);
+            }
         }
         return newTasks.build();
     }
 
-    private Set<RemoteTask> createTaskOnRandomNode()
+    private Optional<RemoteTask> createTaskOnRandomNode()
     {
-        checkState(stage.getScheduledNodes().isEmpty(), "Stage task is already scheduled on node");
+        checkState(scheduledTasks.isEmpty(), "Stage task is already scheduled on node");
         List<InternalNode> allNodes = splitPlacementPolicy.allNodes();
         checkState(allNodes.size() > 0, "No nodes available");
         InternalNode node = allNodes.get(ThreadLocalRandom.current().nextInt(0, allNodes.size()));
-        return stage.scheduleSplits(node, ImmutableMultimap.of(), ImmutableMultimap.of());
+        return scheduleTask(node, ImmutableMultimap.of(), ImmutableMultimap.of());
     }
 
     private Set<RemoteTask> finalizeTaskCreationIfNecessary()
     {
         // only lock down tasks if there is a sub stage that could block waiting for this stage to create all tasks
-        if (stage.getFragment().isLeaf()) {
+        if (stageExecution.getFragment().isLeaf()) {
             return ImmutableSet.of();
         }
 
         splitPlacementPolicy.lockDownNodes();
 
-        Set<InternalNode> scheduledNodes = stage.getScheduledNodes();
-        Set<RemoteTask> newTasks = splitPlacementPolicy.allNodes().stream()
-                .filter(node -> !scheduledNodes.contains(node))
-                .flatMap(node -> stage.scheduleSplits(node, ImmutableMultimap.of(), ImmutableMultimap.of()).stream())
-                .collect(toImmutableSet());
+        ImmutableSet.Builder<RemoteTask> newTasks = ImmutableSet.builder();
```

Comment:
This can be done side-effect free:

```java
Set<RemoteTask> newTasks = splitPlacementPolicy.allNodes().stream()
        .filter(node -> !scheduledTasks.containsKey(node))
        .map(node -> scheduleTask(node, ImmutableMultimap.of(), ImmutableMultimap.of()))
        .filter(Optional::isPresent)
        .map(Optional::get)
        .collect(toImmutableSet());
```